File: q1160/q1160.java
class Solution {
    public int countCharacters(String[] words, String chars) {
        Map<Character, Integer> charCount = new HashMap<>();
        for (char c : chars.toCharArray()) {
            charCount.put(c, charCount.getOrDefault(c, 0) + 1);
        }

        int res = 0;
        for (String word : words) {
            if (canForm(word, charCount))
                res += word.length();
        }

        return res;
    }

    private boolean canForm(String word, Map<Character, Integer> charCount) {
        Map<Character, Integer> wordCharCount = new HashMap<>();
        for (char c : word.toCharArray()) {
            wordCharCount.put(c, wordCharCount.getOrDefault(c, 0) + 1);
        }

        for (Character c : wordCharCount.keySet()) {
            if (!charCount.containsKey(c) ||
                    charCount.get(c) < wordCharCount.get(c))
                return false;
        }

        return true;
    }
}

File: q1160/q1160.cpp
class Solution
{
public:
    int countCharacters(vector<string> &words, string chars)
    {
        unordered_map<char, int> charCount;
        for (char c : chars)
            charCount[c]++;

        int res = 0;
        for (const string &word : words)
        {
            if (canForm(word, charCount))
                res += word.size();
        }

        return res;
    }

private:
    bool canForm(const string &word, unordered_map<char, int> &charCount)
    {
        unordered_map<char, int> wordCharCount;
        for (char c : word)
            wordCharCount[c]++;

        for (const auto &[c, count] : wordCharCount)
        {
            if (charCount[c] < count)
                return false;
        }
        return true;
    }
};